cpp/CMakeLists.txt

```diff
@@ -735,6 +735,7 @@ endif()
 
 if(ARROW_WITH_UTF8PROC)
   list(APPEND ARROW_LINK_LIBS utf8proc::utf8proc)
+  list(APPEND ARROW_LINK_LIBS re2)
```

Comment:
That's the correct approach but maybe outside of the `ARROW_WITH_UTF8PROC` scope as we would always link `re2` independently of the `utf8proc`-based kernels?

Reply:
So you say re2 should be a hard dependency now? I'm ok with that.

Reply:
We might need an `ARROW_WITH_RE2` CMake option in keeping with our minimal build configuration mantra

Reply:
I guess that also means it is better we have kernels that refer to RE2 being used as I did in #8468 (replace_substring and replace_substring_re2) instead of having a kernel that will bail out at runtime because re2 was not enabled.

Reply:
Rather than opting out of dependencies, would it make more sense to have an `ARROW_COMPUTE_STRINGS` (feel free to pick a better name) feature flag that, if enabled, requires utf8proc and re2? Unlike the compression libraries that have the `ARROW_WITH_X` flags, these dependencies are tied to specific functionality that can't really exist without them, right?

Reply:
Not all of them need the libraries, all the ASCII kernels run without. But for simplicity, I don't think it's a bad idea. 

Could it be an option that enabling ARROW_COMPUTE_STRINGS by default enables ARROW_WITH_RE2 and ARROW_WITH_UTF8PROC, but still allow you to override ARROW_WITH_RE2/ARROW_WITH_UTF8PROC. This would allow someone to build with just the ASCII kernels, or just ASCII and ARROW_WITH_UTF8PROC based kernels.